cts/cts-regression.in

```diff
@@ -1,230 +1,335 @@
+
+    This class handles special cases and cleans up the `components`
+    list. Specifically, it does the following:
+      * Enforce a default value of ['cli', 'scheduler'].
+      * Replace the 'all' alias with the components that it represents.
+      * Get rid of duplicates.
+
+    The main motivation is that when the `choices` argument of
+    :meth:`parser.add_argument()` is specified, the `default` argument
+    must contain exactly one value (not `None` and not a list). We want
+    our default to be a list of components, namely `cli` and
+    `scheduler`.
+    """
+
+    def __call__(self, parser, namespace, values, option_string=None):
+        all_components = ['cli', 'exec', 'fencing', 'scheduler']
+        default_components = ['cli', 'scheduler']
+
+        if not values:
+            setattr(namespace, self.dest, default_components)
+            return
+
+        # If no argument is specified, the default gets passed as a
+        # string 'default' instead of as a list ['default']. Probably
+        # a bug in argparse. The below gives us a list.
+        if not isinstance(values, list):
+            values = [values]
+
+        components = set(values)
+
+        # If 'all', is found, replace it with the components it represents.
+        try:
+            components.remove('all')
+            components.update(set(all_components))
+        except KeyError:
+            pass
+
+        # Same for 'default'
+        try:
+            components.remove('default')
+            components.update(set(default_components))
+        except KeyError:
+            pass
+
+        setattr(namespace, self.dest, sorted(list(components)))
+
+
+def error_print(msg):
+    """Print an error message.
+
+    :param msg: Message to print.
+    :type msg: str
+    """
+    print('      * ERROR:   %s' % msg)
+
+
+def run_components(components, verbose=False, valgrind=False):
+    """Run components' regression tests and report results for each.
+
+    :param components: A list of names of components for which to run
+        tests.
+    :type tests: list[str]
+    :return: :attr:`CrmExit.OK` if all tests were successful,
+        :attr:`CrmExit.ERROR` otherwise.
+    :rtype: :class:`CrmExit`
+    """
+    failed = []
+
+    for comp in components:
+        rc = comp.run(verbose, valgrind)
+        if rc != CrmExit.OK:
+            error_print('%s regression tests failed (%s)' % (comp.name, rc))
+            failed.append(comp.name)
+
+    if failed:
+        print('Failed regression tests:', end='')
+        for comp in failed:
+            print(' %s' % comp, end='')
+        print()
+        return CrmExit.ERROR
+
+    return CrmExit.OK
+
+
+def main():
+    """Run Pacemaker regression tests as specified by arguments."""
+    global TEST_HOME
+    try:
+        TEST_HOME = os.path.dirname(os.readlink(sys.argv[0]))
+    except OSError:
+        TEST_HOME = os.path.dirname(sys.argv[0])
+
+    # Available components
+    components = {
```

Comment:
Could optionally make a helper function (only called once, with hard-coded values plus the remote check) to construct this dict, so that `main()` is shorter.